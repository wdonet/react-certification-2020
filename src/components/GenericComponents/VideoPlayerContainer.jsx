import React, { useEffect, useState } from 'react';
import { useLocation } from 'react-router';
import styled from 'styled-components';
import { Button } from '../../ui';
import RelatedVideosList from './RelatedVideosList';

const FAVORITES_KEY = "favorites";
const getParsedFavorites = () => JSON.parse(window.localStorage.getItem(FAVORITES_KEY));

const StyledVideoPlayerContainer = styled.div`
  display: flex;
`;

const ControlsContainer = styled.div`
  justify-content: flex-end;
  display: flex;
`;

const StyledPlayerSection = styled.div`
  width: 100%;
  align-content: flex-start;
  display: grid;
`;

const VideoPlayerContainer = ({ videosList, onCaptionClick }) => {
  const location = useLocation();
  const { state, search } = location;
  const videoId = new URLSearchParams(search).get("id");

  const getLabel = (videoId) => 
    getParsedFavorites() && getParsedFavorites()[videoId] 
    ? "Remove favorite"
    : "Add favorite";

  const [favoriteButtonLabel, setFavoriteButtonLabel] = useState(getLabel());

  const updateButtonLabel = () => { setFavoriteButtonLabel(() => getLabel(videoId)); }

  useEffect(() => updateButtonLabel());

  useEffect(() => {
    /* global YT */
    /* eslint no-undef: "error" */
    YT.ready(() => {
      new YT.Player('player', {
        height: '460',
        width: '100%',
        videoId,
        events: {
          'onReady': (event) => event.target.playVideo(),
        }
      });
    });
    return () => {
      let player = document.querySelector("#player");
      let newPlayer = document.createElement("div");
      newPlayer.id = "player";
      player.parentNode.replaceChild(newPlayer, player);
      window.YTPlayer = undefined;
    }
  }, [videoId]);

  const addRemoveFavorite = (video, videoId, callback) => {
    let favorites = getParsedFavorites();
    if(!favorites){
      favorites = {[videoId]: video};
    } else if(!favorites[videoId]) {
      favorites[videoId] = video;
    } else {
      delete favorites[videoId];
    }
    favorites = JSON.stringify(favorites);
    window.localStorage.setItem(FAVORITES_KEY, favorites);
    callback && callback();
  }

  return (
    <StyledVideoPlayerContainer id="video-player-container">
      <StyledPlayerSection>
        <div id="player" />
        <ControlsContainer>
          <Button 
              data-testid="add-favorite"
              margin="4px"
              height="30px"
              width="150px"
              onClick={() => addRemoveFavorite(state, videoId, updateButtonLabel)}
          >
            { favoriteButtonLabel }
          </Button>
        </ControlsContainer>
      </StyledPlayerSection>
      <RelatedVideosList 
        videosList={videosList} 
        onCaptionClick={onCaptionClick} 
        addRemoveFavorite={
          (video, callback) => addRemoveFavorite(video, video.id.videoId, callback)
        }
      />
    </StyledVideoPlayerContainer>
  );
};

export default VideoPlayerContainer;
