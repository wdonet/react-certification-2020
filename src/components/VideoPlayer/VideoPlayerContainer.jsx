import React, { useContext, useEffect } from 'react';
import styled from 'styled-components';
import SearchContext from '../../providers/SearchContext';
import RelatedVideoList from "./RelatedVideoList";

const StyledVideoPlayerContainer = styled.div`
  display: flex;
`;

const VideoPlayerContainer = ({ videoId }) => {

  const { videos } = useContext(SearchContext);
  let player = {};

  useEffect(() => {
    let done = false;
    /* global YT */
    /* eslint no-undef: "error" */
    player = new YT.Player('player', {
      height: 0.5 * Math.max(document.documentElement.clientHeight || 0, window.innerHeight || 0),
      width: '60%',
      videoId,
      events: {
        onReady(event) {
          event.target.playVideo();
        },
        onStateChange(event) {
          if (event.data === YT.PlayerState.PLAYING && !done) {
            setTimeout(player.stopVideo, 6000);
            done = true;
          }
        },
      },
    });
  });

  return (
    <StyledVideoPlayerContainer id="video-player-container">
      <div id="player" />
      <RelatedVideoList videos={videos} playVideoByID={(videoID) => player.loadVideoById(videoID,0)}/>
    </StyledVideoPlayerContainer>
  );
};

export default VideoPlayerContainer;
