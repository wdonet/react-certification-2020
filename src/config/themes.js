export default {
  dark: {
    icon: 'moon',
    sound: 'https://www.joshwcomeau.com/sounds/switch-on.mp3',
    background: {
      primary: '#161f23',
      secondary: '#1e2931',
    },
    decorator: {
      primary: '#2c0e7b',
      secondary: '#194c7c',
      contrast: '#FFF',
    },
    text: {
      primary: '#FFF',
      secondary: '#FFF',
    },
  },
  light: {
    icon: 'sun',
    sound: 'https://www.joshwcomeau.com/sounds/switch-off.mp3',
    background: {
      primary: '#eef4f4',
      secondary: '#d7dee7',
    },
    decorator: {
      primary: '#5b8cbd',
      secondary: '#d29b3b',
      contrast: '#222',
    },
    text: {
      primary: '#222',
      secondary: '#222',
    },
  },
}
