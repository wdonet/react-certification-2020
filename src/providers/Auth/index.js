export { default, AuthContext } from './AuthProvider';
export { default as useAuth } from './useAuth';
