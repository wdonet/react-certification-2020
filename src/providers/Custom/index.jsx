import React, { useState } from 'react'
import PropTypes from 'prop-types'
import customs from '../../config/customs'

export const CustomContext = React.createContext({})

const CustomProvider = ({ children }) => {
  const [triggered, setTriggered] = useState({})
  const [themeClicks, setThemeClicks] = useState(0)

  if (themeClicks >= 10) {
    setTriggered({
      ...triggered,
      rainbow: {
        values: customs.rainbow,
        activated: true,
      },
    })
    setThemeClicks(0)
  }

  const increaseThemeClicks = () => {
    setThemeClicks(themeClicks + 1)
  }

  const contextValue = {
    triggered,
    increaseThemeClicks,
  }

  return <CustomContext.Provider value={contextValue}>{children}</CustomContext.Provider>
}

CustomProvider.propTypes = {
  children: PropTypes.node.isRequired,
}

export default CustomProvider
