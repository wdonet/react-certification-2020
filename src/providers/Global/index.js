export { default, GlobalContext } from './GlobalProvider';
export { default as useGlobal } from './useGlobal';
