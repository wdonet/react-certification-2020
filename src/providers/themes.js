const defaultColors = {
  white: '#FFFFFF',
  black: '#000000',
};

const font = {
  h1: {
    fontSize: '',
    fontWeight: '',
    letterSpacing: 1.5,
  },
};

export const lightTheme = {
  font,
  color: {
    ...defaultColors,

    primary: '#dce775',
    primaryVariant: '#c0ca33',
    secondary: '#c8ff00',

    background: '#ffffff',
    surface: '#f5f5f5',
    error: '#e64a19',
  },
};

export const darkTheme = {
  font,
  color: {
    ...defaultColors,

    primary: '#00695c',
    primaryVariant: '#004d40',
    secondary: '#1de9b6',

    background: '#616161',
    surface: '#757575',
    error: '#e65100',
  },
};
