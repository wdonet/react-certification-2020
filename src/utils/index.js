export { default as contextWrapper } from './testing/helpers/contextWrapper';
export { default as renderWithTheme } from './testing/helpers/renderWithTheme';
export { default as googleMockedAPIObject } from './testing/mocks/googleMockedAPIObject';
export { default as youtubeMockedData } from './testing/mocks/youtubeMockedData';
export { default as YTMockedObject } from './testing/mocks/YTMockedObject';
