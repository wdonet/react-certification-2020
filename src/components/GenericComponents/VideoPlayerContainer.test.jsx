import React from 'react';
import 'jest-styled-components';
import { lightTheme } from '../../providers/themes';
import { act, getAllByTestId, render } from '@testing-library/react';
import { YTMockedObject, contextWrapper, youtubeMockedData, routerWrapper } from '../../utils';
import VideoPlayerContainer from './VideoPlayerContainer';
import AppContext from '../../providers/AppContext';

global.YT = YTMockedObject;
const EXPECTED_LENGHT = 3
const contextValueVideosList = youtubeMockedData.items.slice(0, EXPECTED_LENGHT);
const build = async (
  Component = <VideoPlayerContainer videosList={contextValueVideosList}/>, 
  currentVideoId = contextValueVideosList[0].id.videoId
  ) => {
  const contextValue = { search: jest.fn, currentVideoId, theme: lightTheme };
  let container;
  let routeWrap; 
  await act(async () => {
    let contextWrap = contextWrapper(AppContext, contextValue, Component);
    routeWrap = await routerWrapper(contextWrap);
    container = render(routeWrap.wrap).container;
  });

  return { 
    container,
    history: () => routeWrap.history,
    videoCaptionsList: () => getAllByTestId(container, (id) => id.includes('small-caption-')) 
  };
};

describe('VideoPlayerContainer', () => {
  it('shows div for iframe load', async () => {
    const built = await build();
    const { firstChild } = built.container;
    expect(firstChild).toMatchSnapshot();
  });
});
