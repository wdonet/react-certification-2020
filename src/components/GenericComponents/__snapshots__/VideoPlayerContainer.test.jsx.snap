// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPlayerContainer shows div for iframe load 1`] = `
<div
  class="sc-jzJRlG dPOBZH"
  id="video-player-container"
>
  <div
    class="sc-kAzzGY jhRLeD"
  >
    <div
      id="player"
    />
    <div
      class="sc-cSHVUG jTEtdD"
    >
      <button
        class="sc-iwsKbI ekEwku"
        data-testid="add-favorite"
        height="30px"
        width="150px"
      >
        Add favorite
      </button>
    </div>
  </div>
  <div
    class="sc-fjdhpX cjBnfF"
    id="related-videos-list"
  >
    <div
      data-testid="small-caption-{\\"kind\\":\\"youtube#video\\",\\"videoId\\":\\"nmXMgqjQzls\\"}"
    >
      <div
        class="sc-gqjmRU gWskoB"
      >
        <img
          alt="Video Tour | Welcome to Wizeline Guadalajara"
          src="https://i.ytimg.com/vi/nmXMgqjQzls/default.jpg"
        />
        <div
          class="sc-jTzLTM dCsEsz"
        >
          <div
            class="sc-VigVT dPTAIb"
          >
            Follow Hector Padilla, Wizeline Director of Engineering, for a lively tour of our office. In 2018, Wizeline opened its stunning new office in Guadalajara, Jalisco, ...
          </div>
          <button
            class="sc-iwsKbI eUpwXN"
            data-testid="caption-add-favorite"
          >
            Add favorite
          </button>
        </div>
      </div>
    </div>
    <div
      data-testid="small-caption-{\\"kind\\":\\"youtube#video\\",\\"videoId\\":\\"HYyRZiwBWc8\\"}"
    >
      <div
        class="sc-gqjmRU gWskoB"
      >
        <img
          alt="Wizeline Guadalajara | Bringing Silicon Valley to Mexico"
          src="https://i.ytimg.com/vi/HYyRZiwBWc8/default.jpg"
        />
        <div
          class="sc-jTzLTM dCsEsz"
        >
          <div
            class="sc-VigVT dPTAIb"
          >
            Wizeline continues to offer a Silicon Valley culture in burgeoning innovation hubs like Mexico and Vietnam. In 2018, our Guadalajara team moved into a ...
          </div>
          <button
            class="sc-iwsKbI eUpwXN"
            data-testid="caption-add-favorite"
          >
            Add favorite
          </button>
        </div>
      </div>
    </div>
    <div
      data-testid="small-caption-{\\"kind\\":\\"youtube#video\\",\\"videoId\\":\\"Po3VwR_NNGk\\"}"
    >
      <div
        class="sc-gqjmRU gWskoB"
      >
        <img
          alt="Wizeline hace sentir a empleados como en casa"
          src="https://i.ytimg.com/vi/Po3VwR_NNGk/default.jpg"
        />
        <div
          class="sc-jTzLTM dCsEsz"
        >
          <div
            class="sc-VigVT dPTAIb"
          >
            En el 2014, Bismarck fundó Wizeline, compañía tecnológica que trabaja con los corporativos ofreciendo una plataforma para que desarrollen software de forma ...
          </div>
          <button
            class="sc-iwsKbI eUpwXN"
            data-testid="caption-add-favorite"
          >
            Add favorite
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
